routing/integrated_routing_test.go

```diff
@@ -62,11 +65,206 @@ func TestProbabilityExtrapolation(t *testing.T) {
 	// of data from other channels), all ten bad channels will be tried
 	// first before switching to the paid channel.
 	ctx.mcCfg.AprioriWeight = 1
-	attempts, err = ctx.testPayment()
+	attempts, err = ctx.testPayment(1)
 	if err != nil {
 		t.Fatalf("payment failed: %v", err)
 	}
 	if len(attempts) != 11 {
 		t.Fatalf("expected 11 attempts, but needed %v", len(attempts))
 	}
 }
+
+type mppSendTestCase struct {
+	name              string
+	amt               btcutil.Amount
+	expectedAttempts  int
+	expectedSuccesses []expectedHtlcSuccess
+	graph             func(g *mockGraph)
+	expectedFailure   bool
+	maxHtlcs          uint32
+}
+
+func onePathGraph(g *mockGraph) {
+	// Create the following network of nodes:
+	// source -> intermediate1 -> target
+
+	const im1NodeID = 3
+	intermediate1 := newMockNode(im1NodeID)
+	g.addNode(intermediate1)
+
+	g.addChannel(13, sourceNodeID, im1NodeID, 200000)
+	g.addChannel(32, targetNodeID, im1NodeID, 100000)
+}
+
+func twoPathGraph(g *mockGraph) {
+	// Create the following network of nodes:
+	// source -> intermediate1 -> target
+	// source -> intermediate2 -> target
+
+	const im1NodeID = 3
+	intermediate1 := newMockNode(im1NodeID)
+	g.addNode(intermediate1)
+
+	const im2NodeID = 4
+	intermediate2 := newMockNode(im2NodeID)
+	g.addNode(intermediate2)
+
+	g.addChannel(13, sourceNodeID, im1NodeID, 200000)
+	g.addChannel(14, sourceNodeID, im2NodeID, 200000)
+	g.addChannel(32, targetNodeID, im1NodeID, 100000)
+	g.addChannel(42, targetNodeID, im2NodeID, 100000)
+}
+
+var mppTestCases = []mppSendTestCase{
+	// Test a two-path graph with sufficient liquidity. It is expected that
+	// pathfinding will try first try to send the full amount via the two
+	// available routes. When that fails, it will half the amount to 35k sat
+	// and retry. That attempt reaches the target successfully. Then the
+	// same route is tried again. Because the channel only had 50k sat, it
+	// will fail. Finally the second route is tried for 35k and it succeeds
+	// too. Mpp payment complete.
+	{
+
+		name:             "sufficient inbound",
+		graph:            twoPathGraph,
+		amt:              70000,
+		expectedAttempts: 5,
+		expectedSuccesses: []expectedHtlcSuccess{
+			{
+				amt:   35000,
+				chans: []uint64{13, 32},
+			},
+			{
+				amt:   35000,
+				chans: []uint64{14, 42},
+			},
+		},
+		maxHtlcs: 1000,
+	},
+
+	// Test that a cap on the max htlcs makes it impossible to pay.
+	{
+		name:              "no splitting",
+		graph:             twoPathGraph,
+		amt:               70000,
+		expectedAttempts:  2,
+		expectedSuccesses: []expectedHtlcSuccess{},
+		expectedFailure:   true,
+		maxHtlcs:          1,
+	},
+
+	// Test that a payment is split in multiple parts that all use the same
+	// route if the full amount cannot be sent in a single htlc.
```

Comment:
It cannot carry the sum, but the receiver doesn't know that upfront. So it will try to 'bin pack' the channel, to finally come to the conclusion that there isn't sufficient liquidity on the network.

Reply:
Ok, I missed that it eventually fails.

Reply:
Added to comment